### ISSUE TITLE
Custom signing backends

### ISSUE THREAD

**Issue by magik6k** [filecoin-project/lotus#2016]
Lotus Daemon should connect to external processes which would handle signing. This could be done using jsonrpc (http(s) or websockets).

Example setup:

conifg
```toml
[Wallet]
ExternalBackends = ["http://127.0.0.1:3210/myHardwareWalletApi"]
```

api exposed by the external process:
```go
type WalletAPI interface {
  WalletNew(context.Context, crypto.SigType) (address.Address, error)
  WalletHas(context.Context, address.Address) (bool, error)
  WalletList(context.Context) ([]address.Address, error)
  WalletSign(context.Context, address.Address, []byte) (*crypto.Signature, error)
  // nonce keeping done by wallet app
  WalletSignMessage(context.Context, address.Address, *types.Message) (*types.SignedMessage, error)
  WalletVerify(context.Context, address.Address, []byte, *crypto.Signature) bool
  WalletExport(context.Context, address.Address) (*types.KeyInfo, error)
  WalletImport(context.Context, *types.KeyInfo) (address.Address, error)
  WalletDelete(context.Context, address.Address) error
}
```

**Comment by Kubuxu** [comment]
I would separate VRF and block signing into their own thing, so you could rate-limit them separately.

**Comment by Kubuxu** [comment]
Also, we really need domain separation tags for signing.
Especially that VRF currently is an arbitrary sign.

**Comment by Schwartz10** [comment]
Assuming there can only be a single external wallet configured at a time, I think this would be really helpful for safer management of a node's funds, and _potentially_ for custodial wallet / storage applications. I don't see how this feature could be used by non-custodial wallet / storage applications, mainly because a Lotus node still remains closely tied with a single signing entity (even if its outsourced).

In a perfect world, any Lotus node method that requires a signature would support a BYOS (bring your own signature) model in addition to optionally shimming wallet methods on the node itself. This approach would seemingly give applications more flexibility in how they handle signatures and auth. It also allows client apps to easily point to different nodes in case one goes down. I haven't done enough digging to be certain, but I'm guessing this is how ethereum wallets like MetaMask and geth work together.

If the node can be configured with more than 1 external wallet, this feedback changes, but I'd follow up with Q's around how clients are expected to make calls to the node, with a specific wallet in mind + provisioning JWTs to allow clients access over a single wallet.

**Comment by Kubuxu** [comment]
I think it is feasible to have multiple backends connected at once (one of them would be the default for the `default` address).

For normal operations, sending funds, calling methods you can bring your own signed message using `MpoolPush`, you, of course, then have to craft the message payload but there are external libraries in works for that.

----

One of the main aims of this functionality is securing miner's worker key, which is required to be used every epoch period (25s) to perform around 5 signing operations.

**Comment by jleni** [comment]
If you are looking for an external signer, I would definitely avoid HTTP, this should be very much forbidden in my opinion. (Minimum should be HTTPS. For high-throughput uses, most projects use things like grpc).
We have already implemented a complete API that provides most of these functions in filecoin-rs and that can be use today. https://github.com/Zondax/filecoin-rs/tree/dev#filecoin-signing-tools

I am a bit confused about what you call "wallet" here. I see two distinct concepts:
- user interaction
- key management

I may be missing something, but how/why the node itself will initiate signing request? Is this something that an actual node needs?

**Comment by Schwartz10** [comment]
> One of the main aims of this functionality is securing miner's worker key

This makes a lot of sense! I've been thinking about this from a client application // public node hosting (like Infura) perspective. 

> Using `MpoolPush`

I should've clarified that I meant configurable external wallets might not be any _more_ useful (imo) than what already exists in Lotus (`MpoolPush`). Although [last I checked](https://filecoinproject.slack.com/archives/CP50PPW2X/p1589485630327400), `MpoolPush` does not work for non-custodial storage applications. Operating in my own bubble, I thought configuring the node to point to external signers was meant to solve the non-custodial storage app problem. And I don't think configuring multiple external signers would solve that problem either (a `default` wallet doesn't really matter to a node hosting provider supporting more than one user).

**Comment by jleni** [comment]
Ok, I didn't read that you want this to secure miner's keys. 

That makes more sense, though, I would expect any serious miners not to run in-process in Lotus. Instead, they very like setup a mining pool backend that can connect to a lotus node to relay signed txs and/or its own remote signer. This is a similar model as exchanges and what we aimed at in the filecoin-rs service.

In any case, I think the signing service should already work well in this scenario. (both soft&hardware based signing). About key storage, the situation is a bit different and may require a tiny bit more work to ensure proper secure storage. 

As signing requests seem to be initiated by the node (which you seem to assume is not secure enough), there are several aspects that should be taking into account such as whitelisting, signing rate-limits, both authentication&authorization, etc. In my opinion, unless really necessary, I would suggest it is better that a node directly exposed to the P2P network should not initiate signing requests itself.

**Comment by popra** [comment]
I think there's quite some interest regarding @Schwartz10 's description of a BYOS (bring your own signature) model for non custodial apps, so I'll try to revive the topic (apologies, if this is not the best issue for the topic).

Currently `MpoolPush` allows rpc connected apps (for example apps running in the browser vs. dialable services) to provide end users with the ability to transfer FIL around, without providing the apps or lotus with their keys. The same ability does not exist, afaik, for storage deals for example. For storage deals, the end user needs to either bring their own lotus, or entrust their keys to some service - this raises the bar of entry into the ecosystem for end users.
 
By no stretch of imagination am I fluent in lotus, but it looks like this might be possible by having lotus ask the ws connected app to sign stuff, here:

https://github.com/filecoin-project/go-fil-markets/blob/adf60de9bc6f39531ca611cd2fbd5c032a8f2060/storagemarket/nodes.go#L52
https://github.com/filecoin-project/go-fil-markets/blob/adf60de9bc6f39531ca611cd2fbd5c032a8f2060/storagemarket/nodes.go#L108

In terms of flow:

- the wc connected rpc client issues a start deal, signaling their intent to handle signing
- they get subscribed (or subscribe) to a "signature request" channel on the ws
- for every "signature request" notification on the ws channel they have to call into a rpc method providing the signature, "signature request", etc.

This would not be less secure than the existing `MpoolPush`.

**Comment by jennijuju** [comment]
closed by #3583
